File: clients/geomqtt-core/src/coord.ts
/**
 * Slippy-map XYZ tile math (Web Mercator). Mirrors the server-side `coord.rs`.
 */

import type { TileCoord } from "./types.js";

const MAX_LAT = 85.05112878;

export function tileForCoord(z: number, lat: number, lon: number): { x: number; y: number } {
  const n = 2 ** z;
  const clampedLat = Math.max(-MAX_LAT, Math.min(MAX_LAT, lat));
  const latRad = (clampedLat * Math.PI) / 180;
  const x = Math.floor(((lon + 180) / 360) * n);
  const y = Math.floor(
    ((1 - Math.log(Math.tan(latRad) + 1 / Math.cos(latRad)) / Math.PI) / 2) * n,
  );
  const max = n - 1;
  return {
    x: Math.max(0, Math.min(max, x)),
    y: Math.max(0, Math.min(max, y)),
  };
}

export function bboxForTile(
  z: number,
  x: number,
  y: number,
): { w: number; s: number; e: number; n: number } {
  const n = 2 ** z;
  const w = (x / n) * 360 - 180;
  const e = ((x + 1) / n) * 360 - 180;
  const north = (Math.atan(Math.sinh(Math.PI * (1 - (2 * y) / n))) * 180) / Math.PI;
  const south = (Math.atan(Math.sinh(Math.PI * (1 - (2 * (y + 1)) / n))) * 180) / Math.PI;
  return { w, s: south, e, n: north };
}

/**
 * All tiles at zoom `z` that intersect the bbox. Handles antimeridian by not
 * wrapping — callers can split a crossed viewport before calling.
 */
export function tilesCoveringBbox(
  z: number,
  w: number,
  s: number,
  e: number,
  n: number,
): TileCoord[] {
  const tl = tileForCoord(z, n, w);
  const br = tileForCoord(z, s, e);
  const xMin = Math.min(tl.x, br.x);
  const xMax = Math.max(tl.x, br.x);
  const yMin = Math.min(tl.y, br.y);
  const yMax = Math.max(tl.y, br.y);
  const out: TileCoord[] = [];
  for (let y = yMin; y <= yMax; y++) {
    for (let x = xMin; x <= xMax; x++) {
      out.push({ z, x, y });
    }
  }
  return out;
}

/**
 * Pick the largest published zoom ≤ `current`. Falls back to the lowest
 * published if `current` is below all of them. Assumes `published` is
 * non-empty and sorted ascending.
 */
export function closestPublishedZoom(current: number, published: number[]): number {
  if (published.length === 0) return Math.floor(current);
  const sorted = [...published].sort((a, b) => a - b);
  const first = sorted[0]!;
  const last = sorted[sorted.length - 1]!;
  if (current <= first) return first;
  if (current >= last) return last;
  let chosen = first;
  for (const z of sorted) {
    if (z <= current) chosen = z;
    else break;
  }
  return chosen;
}

File: crates/geomqtt-server/src/coord.rs
//! Slippy-map XYZ tile math (Web Mercator).

use std::f64::consts::PI;

/// Tile (x, y) at zoom `z` containing `(lat, lon)` in degrees.
pub fn tile_for_coord(z: u8, lat: f64, lon: f64) -> (u32, u32) {
    let n = 2f64.powi(z as i32);
    let x = ((lon + 180.0) / 360.0 * n).floor().clamp(0.0, n - 1.0) as u32;
    let lat_rad = lat.clamp(-85.05112878, 85.05112878).to_radians();
    let y = ((1.0 - (lat_rad.tan() + 1.0 / lat_rad.cos()).ln() / PI) / 2.0 * n)
        .floor()
        .clamp(0.0, n - 1.0) as u32;
    (x, y)
}

/// (west, south, east, north) bbox of tile (z, x, y), degrees.
pub fn bbox_for_tile(z: u8, x: u32, y: u32) -> (f64, f64, f64, f64) {
    let n = 2f64.powi(z as i32);
    let w = x as f64 / n * 360.0 - 180.0;
    let e = (x + 1) as f64 / n * 360.0 - 180.0;
    let north = (PI * (1.0 - 2.0 * y as f64 / n)).sinh().atan().to_degrees();
    let south = (PI * (1.0 - 2.0 * (y + 1) as f64 / n))
        .sinh()
        .atan()
        .to_degrees();
    (w, south, e, north)
}

/// Containing tiles at each of the configured zoom levels for a point.
pub fn tiles_for_point(zooms: &[u8], lat: f64, lon: f64) -> Vec<(u8, u32, u32)> {
    zooms
        .iter()
        .map(|&z| {
            let (x, y) = tile_for_coord(z, lat, lon);
            (z, x, y)
        })
        .collect()
}

/// All tile (x, y) pairs at zoom `z` that intersect the bbox (w, s, e, n).
#[allow(dead_code)]
pub fn tiles_covering_bbox(z: u8, w: f64, s: f64, e: f64, n: f64) -> Vec<(u32, u32)> {
    let (x0, y0) = tile_for_coord(z, n, w);
    let (x1, y1) = tile_for_coord(z, s, e);
    let (x_min, x_max) = (x0.min(x1), x0.max(x1));
    let (y_min, y_max) = (y0.min(y1), y0.max(y1));
    let mut out = Vec::with_capacity(((x_max - x_min + 1) * (y_max - y_min + 1)) as usize);
    for y in y_min..=y_max {
        for x in x_min..=x_max {
            out.push((x, y));
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn zero_zero_at_z0() {
        assert_eq!(tile_for_coord(0, 0.0, 0.0), (0, 0));
    }

    #[test]
    fn roundtrip_bbox_contains_center() {
        let (z, x, y) = (10, 523, 391);
        let (w, s, e, n) = bbox_for_tile(z, x, y);
        let (cx, cy) = ((w + e) / 2.0, (s + n) / 2.0);
        assert_eq!(tile_for_coord(z, cy, cx), (x, y));
    }

    #[test]
    fn tiles_for_point_one_per_zoom() {
        let tiles = tiles_for_point(&[4, 8, 12], 44.49, 11.34);
        assert_eq!(tiles.len(), 3);
        assert_eq!(tiles[0].0, 4);
    }
}
